State management (#70)
deity-io/falcon-docs: docs/falcon-client/state-management.md

title: State Management
---

Falcon comes with [Apollo Client](https://www.apollographql.com/docs/react/v2.5/api/apollo-client/) for client-side state management when you use [`create-falcon-app`](getting-started/installation.md#create-falcon-app) to generate a new application. Apollo Client will be pre-configured and does not require any additional setup. You can start using it right away.

## The short version

> If you already know how to use Apollo Client, you pretty much already know how to use it with Falcon. If you need a refresher, continue to the [long version](falcon-client/state-management.md#the-long-version) or the [Apollo documentation](https://www.apollographql.com/docs/react/v2.5/essentials/get-started/).

Reviewer: ```suggestion
If you already know how to use Apollo Client, you pretty much know how to use it with Falcon. If you need a refresher, continue to the [long version](falcon-client/state-management.md#the-long-version) or the [Apollo documentation](https://www.apollographql.com/docs/react/v2.5/essentials/get-started/).
```

### Fetching and sending data from and to the server

Falcon utilizes Apollo's standard `Query` and `Mutation` components.

### Local state management

Your initialized local state and the resolvers go in `client/src/clientState.js`.

> ### Falcon's additional components

Reviewer: Maybe you can also write something about Falcon's `Query` component (which extends Apollo's `Query`). The main features are displaying error and loading states. Unfortunately this is only useful when using the render prop method. But since this article uses that everywhere, it's useful to know about these features.

Author: @nimobeeren good suggestion, I'll write a bit about that as well.

Falcon comes with a few pre-defined queries and mutations when you create a project with [`create-falcon-app`](getting-started/installation.md#create-falcon-app). These are primarily from our [`@deity/falcon-ecommerce-uikit`](https://github.com/deity-io/falcon/tree/master/packages/falcon-ecommerce-uikit) package and can be of great use when starting out with an eCommerce application with an integrated blog.

> A pattern you will see quite a bit in our included (but optional) components is to create a custom `Query` component with the query added as a default prop.

Author: @Borales I'll try to rephrase it. It is kind of vague, I agree.

```jsx
export class CustomerQuery extends Query {
  static defaultProps = {
    query: GET_CUSTOMER
  };
}
```

## The long version

> This will go in-depth about how to use Apollo for managing the client state of your Falcon application with Apollo.

Reviewer: ```suggestion
This will go in-depth about how to use Apollo for managing the client state of your Falcon application.
```

Author: @nimobeeren I don't think that part should be removed, since most of it is about working with Apollo.

Reviewer: I agree about mentioning Apollo, but it is being mentioned two times in one sentence

Author: oh, I totally missed that. Never mind :sweat_smile:

> If youre unfamiliar with GraphQL queries it is recommended to check out [this guide](https://graphql.org/learn/queries/).

Reviewer 2: ```suggestion
If you're unfamiliar with GraphQL queries it is recommended to check out [this guide](https://graphql.org/learn/queries/).
```

### Fetching data (Query component)

The `Query` API is pretty simple if you're familiar with React and the [Render Prop](https://reactjs.org/docs/render-props.html) pattern.

> React will call the render prop function you provide with an object from Apollo Client containing loading, error, and data properties that you can use to render your UI.

Reviewer 2: ```suggestion
React will call the render prop function you provide with an object from Apollo Client containing `loading`, `error`, and `data` properties that you can use to render your UI.
```

```jsx
import gql from "graphql-tag";
import { Query } from "react-apollo";

const GET_DOGS = gql`
  {
    dogs {
      id
      breed
    }
  }
`;

const Dogs = ({ onDogSelected }) => (
  <Query query={GET_DOGS}>
    {({ loading, error, data }) => {
      if (loading) return "Loading...";
      if (error) return `Error! ${error.message}`;

      return (
        <select name="dog" onChange={onDogSelected}>
          {data.dogs.map(dog => (
            <option key={dog.id} value={dog.breed}>
              {dog.breed}
            </option>
          ))}
        </select>
      );
    }}
  </Query>
);
```

> The `Query` component subscribes to the data through an observable. The result is first checked in the Apollo Client cache. If the result isn't in the cache it will make a request to the server. When the data is received from the server it will be stored in the cache.

Reviewer: ```suggestion
The `Query` component subscribes to the data through an observable. The result is first checked in the Apollo Client cache. If the result isn't in the cache, it will make a request to the server. When the data is received from the server, it will be stored in the cache.
```

If you need near real-time data you can read more about that in the [Apollo documentation](https://www.apollographql.com/docs/react/v2.5/essentials/queries/#polling-and-refetching).

> If you need to notify the user when you are refetching data or if their network status has changes you can read more about that in the [Apollo documentation](https://www.apollographql.com/docs/react/v2.5/essentials/queries/#loading-and-error-state).

Reviewer: ```suggestion
If you need to notify the user when you are refetching data or if their network status has changes, you can read more about that in the [Apollo documentation](https://www.apollographql.com/docs/react/v2.5/essentials/queries/#loading-and-error-state).
```

If you'd like to dive deeper into the Apollo cache's normalization strategy, check out their [advanced caching guide](https://www.apollographql.com/docs/react/v2.5/advanced/caching/).

```jsx
const UPDATE_TODO = gql`
  mutation UpdateTodo($id: String!, $type: String!) {
    updateTodo(id: $id, type: $type) {
      id
      type
    }
  }
`;
```

#### Loading and error states

In the render prop function, we can destructure `loading` and `error` properties off the mutation result. The `Mutation` component also has `onCompleted` and `onError` props in case you would like to provide callbacks instead. Additionally, the mutation result object also has a `called` boolean that tracks whether or not the mutate function has been called.

### Local state management

Apollo is great when using GraphQL to fetch data in your front-end, but you can also use Apollo Client to do local state management, just like you would use the [React Context API](https://reactjs.org/docs/context.html), MobX or Redux. You can use those other options as well, but using Apollo Client for local state management is ideal because your Apollo cache will be the single source of truth in your application's front-end.

> You can read more about local state management with Apollo Client [here](https://www.apollographql.com/docs/react/v2.5/essentials/local-state/). Since Falcon Client does not support Apollo's hooks yet you should use the examples with Render Prop. You can use the dropdown above their code examples to change the examples from hooks to render props.

Reviewer: Maybe mention that `< 2.0.0` does not support hooks

#### Using and updating local client state

> When using local client state you need to write resolvers (queries and mutations) to access and edit the data. You can read more about that in Apollo's documentation about [local resolvers](https://www.apollographql.com/docs/react/v2.5/essentials/local-state/#local-resolvers).

Reviewer: ```suggestion
When using local client state, you need to write resolvers to query and mutate the data. You can read more about that in Apollo's documentation about [local resolvers](https://www.apollographql.com/docs/react/v2.5/essentials/local-state/#local-resolvers).
```
